tools/lib/traverse.js

```diff
@@ -208,12 +219,17 @@ export class TraverseContext {
    * @return {[chromeTypes.NamedTypeSpec, ...chromeTypes.NamedTypeSpec[]][]}
    */
   expandFunctionParams(spec, id) {
-    // If this is actually a Promise-supporting API, then we call ourselves again to support both
-    // callback and Promise-based versions.
+    // If this function uses "returns_asyc", expand it out and call ourselves again to generate
+    // the valid signatures (either just callback or Promise and callback).
     const expanded = this._maybeExpandFunctionReturnsAsync(spec);
     if (expanded) {
+      if (expanded.withPromise && expanded.withCallback) {
```

Comment:
From what I can tell the only reason for this new case is to avoid calling `expandFunctionParams` if `expanded.withPromise` is undefined.

Could we instead add this to the top of `expandFunctionParams`?

```
if (!spec) {
  return [];
}
```

And then remove this new case?

So basically "if there's no spec, there are no possible expansions".

I think that simplifies things a bit.

Reply:
I don't think that is actually the case since it will return:

```
[...empty array, ...array with one item which is the callback]
```

Spreading with `...` adds items to the array, so it starts with an empty array, adds nothing from the first array, and adds one item from the second one.